File: API/models/Character.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from models.Generico import BaseSchema


class ApiReferenceSchema(BaseModel):
    index: Optional[str] = None
    url: Optional[str] = None
    name: Optional[str] = None

    model_config = ConfigDict(extra="allow")


class CharacterProficiencySchema(BaseModel):
    proficiency: ApiReferenceSchema
    value: Optional[int] = None

    model_config = ConfigDict(extra="allow")


class CharacterSpeedSchema(BaseModel):
    walk: Optional[str] = None
    fly: Optional[str] = None
    swim: Optional[str] = None
    climb: Optional[str] = None
    crawl: Optional[str] = None
    burrow: Optional[str] = None
    hover: Optional[str] = None

    model_config = ConfigDict(extra="allow")


class CharacterCashSchema(BaseModel):
    cp: int = 0
    sp: int = 0
    gp: int = 0
    pp: int = 0

    model_config = ConfigDict(extra="allow")


class CharacterInventorySchema(BaseModel):
    items: List[str] = Field(default_factory=list)
    cash: CharacterCashSchema = Field(default_factory=CharacterCashSchema)

    model_config = ConfigDict(extra="allow")


class CharacterSpellcastingSchema(BaseModel):
    spell_slots: Dict[str, int] = Field(default_factory=dict)
    known_spells: List[str] = Field(default_factory=list)
    prepared_spells: List[str] = Field(default_factory=list)
    spell_save_dc: Optional[int] = None
    spell_attack_bonus: Optional[int] = None

    model_config = ConfigDict(extra="allow")


class CharacterSchema(BaseSchema):
    index: Optional[str] = None
    name: str
    player: Optional[str] = None
    level: int
    character_class: ApiReferenceSchema = Field(alias="class")
    subclass: Optional[ApiReferenceSchema] = None
    race: ApiReferenceSchema
    subrace: Optional[ApiReferenceSchema] = None
    alignment: str
    background: ApiReferenceSchema
    hit_points: int
    hit_dice: str
    hit_points_roll: str
    speed: CharacterSpeedSchema
    strength: int
    dexterity: int
    constitution: int
    intelligence: int
    wisdom: int
    charisma: int
    proficiencies: List[CharacterProficiencySchema] = Field(default_factory=list)
    proficiency_bonus: int
    saving_throws: List[str] = Field(default_factory=list)
    inventory: CharacterInventorySchema = Field(default_factory=CharacterInventorySchema)
    spellcasting: Optional[CharacterSpellcastingSchema] = None
    notes: Optional[str] = None
    campaings: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    image: Optional[str] = None
    url: Optional[str] = None
    meta: Optional[Dict[str, Any]] = Field(default=None, alias="_meta")

    model_config = ConfigDict(
        extra="allow",
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "index": "liora",
                "name": "Liora",
                "player": "Ana_id",
                "level": 5,
                "class": {
                    "index": "cleric",
                    "name": "Cleric"
                },
                "subclass": {
                    "index": "light-domain",
                    "name": "Light Domain"
                },
                "race": {
                    "index": "human",
                    "name": "Human"
                },
                "subrace": {
                    "index": "variant-human",
                    "name": "Variant Human"
                },
                "alignment": "neutral-good",
                "background": {
                    "index": "acolyte",
                    "name": "Acolyte"
                },
                "hit_points": 7,
                "hit_dice": "2d6",
                "hit_points_roll": "1d6",
                "speed": {"walk": "30 ft."},
                "strength": 8,
                "dexterity": 14,
                "constitution": 10,
                "intelligence": 10,
                "wisdom": 8,
                "charisma": 8,
                "proficiencies": [
                    {"proficiency": {"index": "skill-insight", "name": "Skill: Insight"}}
                ],
                "proficiency_bonus": 3,
                "saving_throws": ["wis", "cha"],
                "inventory": {
                    "items": ["battleaxe", "shield"],
                    "cash": {"cp": 0, "sp": 4, "gp": 12, "pp": 0},
                },
                "spellcasting": {
                    "spell_slots": {"1": 4, "2": 3, "3": 2},
                    "known_spells": ["guiding-bolt"],
                    "prepared_spells": ["guiding-bolt"],
                    "spell_save_dc": 15,
                    "spell_attack_bonus": 7,
                },
                "notes": "Seguidora de la Aguja del Alba.",
                "campaings": ["SombraDragon2026_id"],
                "tags": ["pj", "clerigo"],
                "_meta": {
                    "is_deleted": False,
                    "image_url": "/api/images/characters/liora.png",
                    "token_url": "/api/tokens/characters/liora-token.png",
                },
            }
        },
    )

File: APP/src/app/interfaces/Character.ts
export interface ApiReference {
  index: string;
  name: string;
  url?: string;
}

export interface CharacterProficiency {
  proficiency: ApiReference;
  value?: number;
}

export interface CharacterSpeed {
  walk?: string;
  fly?: string;
  swim?: string;
  climb?: string;
  crawl?: string;
  burrow?: string;
  hover?: string;
}

export interface CharacterCash {
  cp: number;
  sp: number;
  ep: number;
  gp: number;
  pp: number;
}

export interface WeaponData {
  damage_dice: string;
  damage_type: string;
  ability: 'strength' | 'dexterity';
}

export interface ArmorData {
  base_ac: number;
  armor_type: 'light' | 'medium' | 'heavy' | 'shield';
  max_dex_bonus?: number;
}

export interface InventoryItem {
  id: string;
  name: string;
  type: 'weapon' | 'armor' | 'item';
  quantity: number;
  weight?: number;
  description?: string;
  state: 'equipped' | 'stored' | 'carried';
  weapon_data?: WeaponData;
  armor_data?: ArmorData;
}

export interface Trait {
  id: string;
  name: string;
  description: string;
  source: 'race' | 'subrace' | 'class' | 'background' | 'custom';
}

export interface SpellEntry {
  index: string;
  name: string;
  level: number;
  school?: string;
  // Datos de combate (rellenados desde el catálogo)
  damage_dice?: string;        // ej: "8d6"
  damage_type?: string;        // ej: "fire"
  attack_type?: 'melee' | 'ranged' | 'save' | null;
  save_type?: string;          // ej: "DEX" (para spells de salvación)
  dc_success?: string;         // ej: "half"
}

export interface SpellSlot {
  total: number;
  used: number;
}

export interface Spellcasting {
  spellcasting_ability?: string;
  spell_save_dc: number;
  spell_attack_bonus: number;
  spell_slots: Record<string, SpellSlot>;
  known_spells: SpellEntry[];
  prepared_spells: SpellEntry[];
}

export interface CharacterInventory {
  items: InventoryItem[];
  cash: CharacterCash;
}

export interface Character {
  _id?: string;
  id?: string;
  index?: string;
  name: string;
  player?: string;
  level: number;
  character_class: ApiReference;
  subclass?: ApiReference;
  race: ApiReference;
  subrace?: ApiReference;
  alignment: string;
  background: ApiReference;
  hit_points: number;
  hit_points_current?: number;
  temp_hp?: number;
  hit_dice: string;
  hit_points_roll?: string;
  speed: CharacterSpeed;
  strength: number;
  dexterity: number;
  constitution: number;
  intelligence: number;
  wisdom: number;
  charisma: number;
  proficiency_bonus: number;
  saving_throws: string[];
  proficiencies: CharacterProficiency[];
  inventory: CharacterInventory;
  spellcasting?: Spellcasting;
  traits: Trait[];
  custom_traits: Trait[];
  // Bio
  notes?: string;
  history?: string;
  image?: string;
  personality_traits?: string[];
  ideals?: string[];
  bonds?: string[];
  flaws?: string[];
}

export interface Skill {
  name: string;
  ability: StatKey;
  index: string;
}

export const SKILLS: Skill[] = [
  { name: 'Acrobatics',     ability: 'dexterity',     index: 'skill-acrobatics' },
  { name: 'Animal Handling',ability: 'wisdom',        index: 'skill-animal-handling' },
  { name: 'Arcana',         ability: 'intelligence',  index: 'skill-arcana' },
  { name: 'Athletics',      ability: 'strength',      index: 'skill-athletics' },
  { name: 'Deception',      ability: 'charisma',      index: 'skill-deception' },
  { name: 'History',        ability: 'intelligence',  index: 'skill-history' },
  { name: 'Insight',        ability: 'wisdom',        index: 'skill-insight' },
  { name: 'Intimidation',   ability: 'charisma',      index: 'skill-intimidation' },
  { name: 'Investigation',  ability: 'intelligence',  index: 'skill-investigation' },
  { name: 'Medicine',       ability: 'wisdom',        index: 'skill-medicine' },
  { name: 'Nature',         ability: 'intelligence',  index: 'skill-nature' },
  { name: 'Perception',     ability: 'wisdom',        index: 'skill-perception' },
  { name: 'Performance',    ability: 'charisma',      index: 'skill-performance' },
  { name: 'Persuasion',     ability: 'charisma',      index: 'skill-persuasion' },
  { name: 'Religion',       ability: 'intelligence',  index: 'skill-religion' },
  { name: 'Sleight of Hand',ability: 'dexterity',     index: 'skill-sleight-of-hand' },
  { name: 'Stealth',        ability: 'dexterity',     index: 'skill-stealth' },
  { name: 'Survival',       ability: 'wisdom',        index: 'skill-survival' },
];

export interface CharacterDraft extends Partial<Character> {
  // HP override desde el wizard
  custom_max_hp?: number;
  // Campos temporales del wizard (no van al modelo final)
  character_class_hit_die?: number;
  race_speed?: number;
  racial_bonuses?: Partial<Record<StatKey, number>>;
  proficiency_names?: string[];
  saving_throw_names?: string[];
  background_traits?: Trait[];
  class_spell_slots?: Record<string, SpellSlot>;
  class_spellcasting_ability?: string;
  selected_personality_traits?: string[];
  selected_ideals?: string[];
  selected_bonds?: string[];
  selected_flaws?: string[];
}

export type StatKey = 'strength' | 'dexterity' | 'constitution' | 'intelligence' | 'wisdom' | 'charisma';

export const STAT_LABELS: Record<StatKey, string> = {
  strength: 'Strength',
  dexterity: 'Dexterity',
  constitution: 'Constitution',
  intelligence: 'Intelligence',
  wisdom: 'Wisdom',
  charisma: 'Charisma',
};

export const STAT_ABBREVIATIONS: Record<StatKey, string> = {
  strength: 'STR',
  dexterity: 'DEX',
  constitution: 'CON',
  intelligence: 'INT',
  wisdom: 'WIS',
  charisma: 'CHA',
};

export const ALIGNMENTS = [
  'Lawful Good',
  'Neutral Good',
  'Chaotic Good',
  'Lawful Neutral',
  'True Neutral',
  'Chaotic Neutral',
  'Lawful Evil',
  'Neutral Evil',
  'Chaotic Evil',
];



export const POINT_BUY_COSTS: Record<number, number> = {
  8: 0, 9: 1, 10: 2, 11: 3, 12: 4, 13: 5, 14: 7, 15: 9,
};

export function statModifier(value: number): number {
  return Math.floor((value - 10) / 2);
}

export function signedModifier(value: number): string {
  const mod = statModifier(value);
  return mod >= 0 ? `+${mod}` : `${mod}`;
}
